### PR TITLE
enh(API): move remote to central request to a specific api repository

### PR DESCRIPTION
**Pull request by sc979**
**/!\ Will be merged after 20.10.0 release**

#MON-6188

## Description

Remove base endpoint path from topology service to information service
Define API version as a constant in the information controller

**Fixes** # (MON-6093)

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.04.x
- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x (master)

### CODE REVIEW

File: src/Centreon/Domain/PlatformTopology/PlatformTopology.php [hunk 1]
```diff
@@ -230,7 +231,10 @@ public function setAddress(?string $address): self
      */
     public function getParentAddress(): ?string
     {
-        return $this->parentAddress;
+        if (null !== $this->parentAddress && $this->getType() === static::TYPE_CENTRAL) {
+            throw new \InvalidArgumentException(_("Cannot use parent address on a Central server type"));
+        }
+        return $this->checkIpAddress($this->parentAddress);
```
Comment (callapa) [inline src/Centreon/Domain/PlatformTopology/PlatformTopology.php:237]: This control must be define in the setter, not getter.

File: src/Centreon/Domain/PlatformTopology/PlatformTopology.php [hunk 2]
```diff
@@ -252,6 +253,9 @@ public function setParentAddress(?string $parentAddress): self
      */
     public function getParentId(): ?int
     {
+        if (null !== $this->parentId && $this->getType() === static::TYPE_CENTRAL) {
```
Comment (callapa) [inline src/Centreon/Domain/PlatformTopology/PlatformTopology.php:256]: This control must be define in the setter, not getter.

File: src/Centreon/Domain/PlatformInformation/PlatformInformation.php [hunk 3]
```diff
@@ -308,6 +308,12 @@ public function setApiPort(?int $port): self
      */
     public function getApiPath(): ?string
     {
+        $this->apiPath = trim(filter_var($this->apiPath, FILTER_SANITIZE_STRING, ['options' => ['default' => '']]), '/');
```
Comment (callapa) [inline src/Centreon/Domain/PlatformInformation/PlatformInformation.php:311]: This control must be define in the setter, not getter.

File: src/Centreon/Domain/PlatformTopology/Interfaces/PlatformTopologyRegisterRepositoryInterface.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+use Centreon\Domain\PlatformInformation\PlatformInformation;
+use Centreon\Domain\PlatformTopology\PlatformTopology;
+use Centreon\Domain\PlatformTopology\PlatformTopologyConflictException;
+use Centreon\Domain\Proxy\Proxy;
+use Centreon\Domain\Repository\RepositoryException;
+
+interface PlatformTopologyRegisterRepositoryInterface
+{
+    /**
+     * @param PlatformTopology $platformTopology
```
Comment (callapa) [inline src/Centreon/Domain/PlatformTopology/Interfaces/PlatformTopologyRegisterRepositoryInterface.php:34]: Don't forget to add a little description

File: src/Centreon/Domain/PlatformTopology/PlatformTopology.php [hunk 5]
```diff
@@ -113,6 +113,22 @@ public function setId(?int $id): self
      */
     public function getType(): ?string
     {
+        if (null === $this->type) {
+            return null;
+        }
+        $this->type = strtolower($this->type);
+
+        // Check if the server_type is available
+        if (!in_array($this->type, static::AVAILABLE_TYPES, true)) {
```
Comment (callapa) [inline src/Centreon/Domain/PlatformTopology/PlatformTopology.php:122]: This control must be define in the setter, not getter.

**Review by callapa — CHANGES_REQUESTED** [review]

File: src/Centreon/Application/ApiPlatform.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        return $this->version;
+    }
+
+    /**
+     * Set the API version
+     *
+     * @param float $version
+     * @return ApiPlatform
+     */
+    public function setVersion(float $version): ApiPlatform
```
Comment (jeremyjaouen) [inline src/Centreon/Application/ApiPlatform.php:51]:
```suggestion
    public function setVersion(float $version): self
```

File: src/Centreon/Application/ApiPlatform.php [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+     * @var float
+     */
+    private $version;
+
+    /**
+     * Get the API version
+     *
+     * @return float
+     */
+    public function getVersion(): float
```
Comment (jeremyjaouen) [inline src/Centreon/Application/ApiPlatform.php:40]: this couldn't be nullable ?

Comment (sc979) [reply src/Centreon/Application/ApiPlatform.php:40]: I don't think so. As the APi version is mandatory and if it's missing a 403 Forbidden is returned

**Review by jeremyjaouen — APPROVED** [review]

**Comment by sc979** [comment]
To be merged after 20.10.0 release.

File: src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php [hunk 8]
```diff
@@ -221,133 +215,20 @@ public function addPlatformToTopology(PlatformTopology $platformTopology): void
...
+                    $platformTopology,
+                    $foundPlatformInformation,
+                    $this->proxyService->getProxy()
                 );
+            } catch (PlatformTopologyConflictException $ex) {
+                throw $ex;
+            } catch (RepositoryException $ex) {
+                throw new PlatformTopologyException($ex->getMessage(), $ex->getCode(), $ex);
+            } catch (\Exception $ex) {
+                throw new PlatformTopologyException('', 0, $ex);
```
Comment (kduret) [inline src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php:231]: not sure it is a good practice to let exception message empty (cc @callapa)

**Review by kduret — DISMISSED** [review]

File: src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php (hunk 8 above)
Comment (sc979) [reply src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php:231]: indeed I forgot this one

Comment (sc979) [reply src/Centreon/Domain/PlatformTopology/PlatformTopologyService.php]: done
conflicts resolved and branch rebased

**Review by kduret — DISMISSED** [review]

**Review by callapa — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**